File: zak/core/audit/logger.py
"""
ZAK Audit Logger — structured audit log emission using structlog.

All audit events are serialized as JSON and tagged with tenant_id + trace_id.
"""

from __future__ import annotations

import sys
from typing import Any

import structlog

from zak.core.audit.events import AuditEvent, AuditEventType


def _configure_structlog() -> None:
    structlog.configure(
        processors=[
            structlog.stdlib.add_log_level,
            structlog.processors.TimeStamper(fmt="iso"),
            structlog.processors.JSONRenderer(),
        ],
        wrapper_class=structlog.BoundLogger,
        context_class=dict,
        logger_factory=structlog.PrintLoggerFactory(file=sys.stdout),
        cache_logger_on_first_use=True,
    )


_configure_structlog()


class AuditLogger:
    """
    Tenant-scoped, agent-scoped structured audit logger.

    Usage:
        logger = AuditLogger(tenant_id="acme", agent_id="risk-quant-v1", trace_id="abc123")
        logger.emit(AgentStartedEvent(tenant_id="acme", agent_id="risk-quant-v1", trace_id="abc123"))
    """

    def __init__(self, tenant_id: str, agent_id: str, trace_id: str) -> None:
        self.tenant_id = tenant_id
        self.agent_id = agent_id
        self.trace_id = trace_id
        self._log = structlog.get_logger().bind(
            tenant_id=tenant_id,
            agent_id=agent_id,
            trace_id=trace_id,
        )

    def emit(self, event: AuditEvent) -> None:
        """Emit a typed audit event as a structured log line."""
        self._log.info(
            event.event_type.value,
            **event.payload,
            timestamp=event.timestamp.isoformat(),
        )

    def log_raw(self, event_type: AuditEventType, **kwargs: Any) -> None:
        """Emit an ad-hoc audit event without a typed model (convenience method)."""
        self._log.info(event_type.value, **kwargs)

File: adk/go/pkg/audit/logger.go
package audit

import (
	"io"
	"os"

	"github.com/rs/zerolog"
)

// Logger is a tenant-scoped, agent-scoped structured audit logger.
//
// Usage:
//
//	logger := audit.NewLogger("acme", "risk-quant-v1", "abc123")
//	logger.Emit(audit.AgentStartedEvent("risk-quant-v1", "acme", "abc123"))
type Logger struct {
	TenantID string
	AgentID  string
	TraceID  string
	log      zerolog.Logger
}

// NewLogger creates a new AuditLogger that writes JSON to stdout.
func NewLogger(tenantID, agentID, traceID string) *Logger {
	return NewLoggerWithWriter(tenantID, agentID, traceID, os.Stdout)
}

// NewLoggerWithWriter creates a new AuditLogger with a custom writer.
func NewLoggerWithWriter(tenantID, agentID, traceID string, w io.Writer) *Logger {
	zl := zerolog.New(w).With().
		Str("tenant_id", tenantID).
		Str("agent_id", agentID).
		Str("trace_id", traceID).
		Logger()

	return &Logger{
		TenantID: tenantID,
		AgentID:  agentID,
		TraceID:  traceID,
		log:      zl,
	}
}

// Emit writes a typed audit event as a structured JSON log line.
func (l *Logger) Emit(event *Event) {
	logEvent := l.log.Info().
		Str("timestamp", event.Timestamp.Format("2006-01-02T15:04:05.000Z07:00"))

	// Include all payload fields
	for k, v := range event.Payload {
		logEvent = logEvent.Interface(k, v)
	}

	logEvent.Msg(string(event.EventType))
}

// LogRaw emits an ad-hoc audit event without a typed model.
func (l *Logger) LogRaw(eventType EventType, fields map[string]interface{}) {
	logEvent := l.log.Info()
	for k, v := range fields {
		logEvent = logEvent.Interface(k, v)
	}
	logEvent.Msg(string(eventType))
}
